src/serac/physics/utilities/functional/functional.hpp

```diff
@@ -135,65 +146,47 @@ class Functional<test(trial)> : public mfem::Operator {
+    auto num_elements = domain.GetNE();
+    SLIC_ERROR_IF(num_elements == 0, "Mesh has no elements");
 
-      auto dim = domain.Dimension();
-      for (int e = 0; e < num_elements; e++) {
-        SLIC_ERROR_IF(domain.GetElementType(e) != supported_types[dim], "Mesh contains unsupported element type");
-      }
+    SLIC_ERROR_IF(dim != domain.Dimension(), "Error: invalid mesh dimension for domain integral");
+    for (int e = 0; e < num_elements; e++) {
+      SLIC_ERROR_IF(domain.GetElementType(e) != supported_types[dim], "Mesh contains unsupported element type");
+    }
 
-      const mfem::FiniteElement&   el = *test_space_->GetFE(0);
-      const mfem::IntegrationRule& ir = mfem::IntRules.Get(el.GetGeomType(), el.GetOrder() * 2);
+    const mfem::FiniteElement&   el = *test_space_->GetFE(0);
+    const mfem::IntegrationRule& ir = mfem::IntRules.Get(el.GetGeomType(), el.GetOrder() * 2);
 
-      constexpr auto flags = mfem::GeometricFactors::COORDINATES | mfem::GeometricFactors::JACOBIANS;
-      auto           geom  = domain.GetGeometricFactors(ir, flags);
-      domain_integrals_.emplace_back(num_elements, geom->J, geom->X, Dimension<geometry_dim>{},
-                                     Dimension<spatial_dim>{}, integrand);
-      return;
-    }
-#ifdef ENABLE_BOUNDARY_INTEGRALS
-    else if constexpr ((geometry_dim + 1) == spatial_dim) {
-      // TODO: fix mfem::FaceGeometricFactors
-      constexpr bool always_false = (geometry_dim == spatial_dim);
-      static_assert(always_false,
-                    "unsupported integral dimensionality due to unimplemented features in mfem::FaceGeometricFactors");
-
-      auto num_boundary_elements = domain.GetNBE();
-      SLIC_ERROR_IF(num_boundary_elements == 0, "Mesh has no boundary elements");
-
-      auto dim = domain.Dimension();
-      SLIC_ERROR_IF(dim != spatial_dim, "Error: invalid mesh dimension for integral");
-      for (int e = 0; e < num_boundary_elements; e++) {
-        SLIC_ERROR_IF(domain.GetBdrElementType(e) != supported_types[dim - 1],
-                      "Mesh contains unsupported element type");
-      }
+    constexpr auto flags = mfem::GeometricFactors::COORDINATES | mfem::GeometricFactors::JACOBIANS;
+    auto           geom  = domain.GetGeometricFactors(ir, flags);
+    domain_integrals_.emplace_back(num_elements, geom->J, geom->X, Dimension<dim>{}, integrand);
+  }
 
-      const mfem::FiniteElement&   el = *test_space_->GetBE(0);
-      const mfem::IntegrationRule& ir = mfem::IntRules.Get(el.GetGeomType(), el.GetOrder() * 2);
-      constexpr auto flags            = mfem::FaceGeometricFactors::COORDINATES | mfem::FaceGeometricFactors::JACOBIANS;
+  template <int dim, typename lambda>
+  void AddBoundaryIntegral(Dimension<dim>, lambda&& integrand, mfem::Mesh& domain)
+  {
+    // TODO: fix mfem::FaceGeometricFactors
+    auto num_boundary_elements = domain.GetNBE();
+    SLIC_ERROR_IF(num_boundary_elements == 0, "Mesh has no boundary elements");
```

Comment:
Good question, I have no idea! It's not clear to me if `mfem::ParMesh::GetNBE()` returns the number of boundary elements on the local mesh, or the entire mesh (in which case it wouldn't be a problem).

Reply:
My understanding is that it will return the information in the local mesh - I don't think any of the `Get*` methods implemented in `mfem::Mesh` are reimplemented in `ParMesh` so they'll all just return info on the local mesh.

Reply:
Some of the parallel "Get" methods in the API have the word `Global` in them.

Reply:
Oops, I just asked Jamie the same question, and got the same answer! Thanks, I'll reimplement this to just be an early exit condition if there are no boundary elements.